src/v/cloud_storage/cache_service.cc

```diff
@@ -9,49 +9,110 @@
  */
 
 #include "cloud_storage/logger.h"
-#include "random/generators.h"
 #include "utils/gate_guard.h"
+#include "vassert.h"
 #include "vlog.h"
 
 #include <seastar/core/coroutine.hh>
-#include <seastar/core/file.hh>
 #include <seastar/core/fstream.hh>
-#include <seastar/core/future.hh>
-#include <seastar/core/print.hh>
 #include <seastar/core/seastar.hh>
 #include <seastar/core/smp.hh>
-#include <seastar/util/defer.hh>
 
 #include <cloud_storage/cache_service.h>
 
 #include <exception>
 #include <filesystem>
-#include <optional>
-#include <string>
-#include <system_error>
 
 namespace cloud_storage {
 
-cache::cache(std::filesystem::path cache_dir) noexcept
+cache::cache(
+  std::filesystem::path cache_dir,
+  size_t max_cache_size,
+  ss::lowres_clock::duration check_period) noexcept
   : _cache_dir(std::move(cache_dir))
-  , _cnt(0) {}
+  , _max_cache_size(max_cache_size)
+  , _check_period(check_period)
+  , _cnt(0)
+  , _total_cleaned(0) {}
+
+uint64_t cache::get_total_cleaned() { return _total_cleaned; }
+
+ss::future<> cache::clean_up_cache() {
+    gate_guard guard{_gate};
+    co_await _walker.walk(_cache_dir.native());
+
+    if (_walker.get_current_cache_size() >= _max_cache_size) {
+        auto size_to_delete = _walker.get_current_cache_size()
+                              - (_max_cache_size * _cache_size_low_watermark);
+
+        auto candidates_for_deletion = _walker.get_files();
+
+        uint64_t deleted_size = 0;
+        size_t i_to_delete = 0;
+        while (deleted_size < size_to_delete) {
+            // todo(Lena): don't delete .part file since they may be in use
+            bool successfully_deleted = true;
+            auto file_to_remove = candidates_for_deletion[i_to_delete].filename;
+            vassert(
+              0
+                == file_to_remove.substr(0, _cache_dir.native().length())
+                     .compare(_cache_dir.native()),
+              "Tried to clean up {}, which is outside of cache_dir {}.",
+              file_to_remove,
+              _cache_dir.native());
+
+            try {
+                co_await ss::remove_file(file_to_remove);
```

Comment:
nit: this will remove data sequentially, one possible optimization here is to collect the names of files in the separate collection `files_to_remove` and then run 

```cpp
co_await ss::max_concurrent_for_each(files_to_remove, 10, [] (std::filesystem::path p) {
    co_await ss::remove_file(p);
});
```
this will delete all files in `files_to_remove` collection but it will run the lambda in parallel with up to 10 parallel fibers

Reply:
Also, before removing the file please check that this is the cache file.

Reply:
Just in case if the directory walker will misbehave. We don't let cache eviction delete some user files.

Reply:
Done.